### PR TITLE
Adding pubsub routing support

### PR DESCRIPTION
**Pull request by pkedy**
# Description

Adding a rule attribute to the topic annotation to utilize PubSub routing.

## Issue reference

dapr/dapr#2582

## Checklist

* [x] Code compiles correctly
* [x] Created/updated tests
* [x] Extended the documentation

### CODE REVIEW

File: sdk-springboot/src/main/java/io/dapr/springboot/DaprSubscriptionBuilder.java [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+    this.rules = new ArrayList<>();
+    this.defaultPath = null;
+    this.metadata = Collections.emptyMap();
+  }
+
+  /**
+   * Sets the default path for the subscription.
+   * @param path The default path.
+   */
+  public void setDefaultPath(String path) {
```
Comment (artursouza) [inline sdk-springboot/src/main/java/io/dapr/springboot/DaprSubscriptionBuilder.java:39]: nit: we return the object back in the setters, to offer fluent API.

File: sdk-springboot/src/main/java/io/dapr/springboot/DaprSubscriptionBuilder.java [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+    defaultPath = path;
+  }
+
+  /**
+   * Adds a rule to the subscription.
+   * @param path The path to route to.
+   * @param match The CEL expression the event must match.
+   * @param priority The priority of the rule.
+   */
+  public void addRule(String path, String match, int priority) {
```
Comment (artursouza) [inline sdk-springboot/src/main/java/io/dapr/springboot/DaprSubscriptionBuilder.java:55]: same.

File: sdk-springboot/src/main/java/io/dapr/springboot/DaprTopicSubscription.java [hunk 3]
```diff
@@ -15,19 +15,23 @@
...
 
   /**
    * Create a subscription topic.
    * @param pubsubName The pubsub name to subscribe to.
    * @param topic The topic to subscribe to.
    * @param route Destination route for messages.
+   * @param routes Destination routes with rules for messages.
    * @param metadata Metdata for extended subscription functionality.
    */
-  public DaprTopicSubscription(String pubsubName, String topic, String route, Map<String, String> metadata) {
```
Comment (artursouza) [inline sdk-springboot/src/main/java/io/dapr/springboot/DaprTopicSubscription.java:27]: This is a breaking change. Please, keep the original constructor and add a new one instead.

File: sdk/src/main/java/io/dapr/Rule.java [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+public @interface Rule {
+  /**
+   * The Common Expression Language (CEL) expression to use
+   * to match the incoming cloud event.
+   * @return the CEL expression.
+   */
+  String match();
+
+  /**
+   * Priority of the rule used for ordering.
```
Comment (artursouza) [inline sdk/src/main/java/io/dapr/Rule.java:26]: Explain here if lower number = higher priority.

File: sdk-springboot/src/main/java/io/dapr/springboot/DaprSubscriptionBuilder.java [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+package io.dapr.springboot;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+public class DaprSubscriptionBuilder {
```
Comment (artursouza) [inline sdk-springboot/src/main/java/io/dapr/springboot/DaprSubscriptionBuilder.java:15]: Does this really need to be public? Make it package visible if possible. It is only used in a private property.

File: sdk-springboot/src/main/java/io/dapr/springboot/DaprTopicKey.java [hunk 6]
```diff
@@ -0,0 +1,43 @@
+/*
+ * Copyright (c) Microsoft Corporation and Dapr Contributors.
+ * Licensed under the MIT License.
+ */
+
+package io.dapr.springboot;
+
+import java.util.Objects;
+
+public class DaprTopicKey {
```
Comment (artursouza) [inline sdk-springboot/src/main/java/io/dapr/springboot/DaprTopicKey.java:10]: Same. Can this be package visible only?

File: sdk-springboot/src/main/java/io/dapr/springboot/DaprTopicRule.java [hunk 7]
```diff
@@ -0,0 +1,43 @@
+/*
+ * Copyright (c) Microsoft Corporation and Dapr Contributors.
+ * Licensed under the MIT License.
+ */
+
+package io.dapr.springboot;
+
+import java.util.Objects;
+
+public class DaprTopicRule {
```
Comment (artursouza) [inline sdk-springboot/src/main/java/io/dapr/springboot/DaprTopicRule.java:10]: Same. Package visible?

**Review by artursouza — CHANGES_REQUESTED** [review]
LGTM overall. Just some minor comments.

**Comment by johnou** [comment]
ping @pkedy

**Comment by artursouza** [comment]
This PR is idle and available to anyone that would like to take over and then create a new PR to replace this one.

File: sdk-springboot/src/main/java/io/dapr/springboot/DaprSubscriptionBuilder.java (hunk 1 above)
Comment (artursouza) [reply sdk-springboot/src/main/java/io/dapr/springboot/DaprSubscriptionBuilder.java:39]: done.

File: sdk-springboot/src/main/java/io/dapr/springboot/DaprSubscriptionBuilder.java (hunk 2 above)
Comment (artursouza) [reply sdk-springboot/src/main/java/io/dapr/springboot/DaprSubscriptionBuilder.java:55]: done.

File: sdk-springboot/src/main/java/io/dapr/springboot/DaprTopicSubscription.java (hunk 3 above)
Comment (artursouza) [reply sdk-springboot/src/main/java/io/dapr/springboot/DaprTopicSubscription.java:35]: Done.

**Review by artursouza — APPROVED** [review]

**Pull request merged**